### PR TITLE
Monitor for FloatingIPs to make the resources reservable

### PR DESCRIPTION
**Pull request by AnishReddyRavula**
make the monitor plugin for FLoating IP plugin - mostly tried to make it similar to host monitor plugin

### CODE REVIEW

**Review by Mark-Powers — COMMENTED** [review]
I think there is some confusion about the different floating IP resources and the monitor interface.

Regarding the monitor:
- `set_reservable()` is automatically called based on the returned `failed/recovered` lists. And those two lists should only contain the fips that changed (e.g. are marked as reservable, but actually look unhealthy) For example, in the [host monitor](https://github.com/ChameleonCloud/blazar/blob/chameleoncloud/xena/blazar/plugins/oshosts/host_plugin.py#L952-L954) it only uses the list of recovered hosts is a sublist of the unreservable hosts.

Regarding fip objects:
- there is the floating ip stored in the blazar db FloatingIp table that you added queries for. It seems like you are using this and the allocations to check if an error occured
- the error the user sees most often with floating ip is "floating ip already allocated in subnet". This is referring to neutron floating ips that should've been cleaned up were not.
- So, I don't think you can fully solve the problem just in blazar. You'll need to query neutron for floating IPs from past reservations, and check if they did not get cleaned up. I think the easiest way to do this is look at the `tags` of a neutron floating ip, when a floating IP lease starts, we tag the floating ip with the reservation id [here](https://github.com/ChameleonCloud/blazar/blob/chameleoncloud/xena/blazar/utils/openstack/neutron.py#L91-L93). If that reservation id is from a reservation that is over, it can safely be deleted.

One way to test this, would be to create a lease for a floating ip, and then while the lease is active inspect it with `openstack floating ip show IP_ADDRESS`. Then, the lease will end, and this will be deleted. If you manually ran `openstack floating ip create` with the same details from `floating ip show`, what we want is for this monitor to delete that ip after it executes.

File: blazar/plugins/floatingips/floatingip_plugin.py [hunk 1]
```diff
@@ -447,3 +455,109 @@ def query_fip_allocations(self, fips, detail=None, lease_id=None,
...
+        def process_fip(fip):
+            fip_address = fip["floating_ip_address"]
+            fip_curr_reservation = db_utils.get_most_recent_reservation_info_by_fip_id(fip['id'])
+            if fip_curr_reservation and fip_curr_reservation['status'] == status.reservation.ACTIVE:
+                # This means that FIP works fine and no need to recover
+                LOG.debug(f"FIP {fip_address} is in active reservation {fip_curr_reservation['id']} - skipping")
+                return
+            fip_pool = neutron.FloatingIPPool(fip['floating_network_id'])
+            # check if the FIP is in neutron subnet allocation pools
+            try:
```
Comment (Mark-Powers) [inline blazar/plugins/floatingips/floatingip_plugin.py:512]: This try block I am not sure about. It seems reasonable, but I don't fully understand all of the expected configuration. @msherman64 could you take a look?

File: blazar/plugins/floatingips/floatingip_plugin.py [hunk 2]
```diff
@@ -447,3 +455,109 @@ def query_fip_allocations(self, fips, detail=None, lease_id=None,
...
+            # if the FIP is not found in any subnet, it can be reserved
+            except utils_exceptions.FloatingIPSubnetNotFound as e:
+                LOG.warning(f"Floating ip {fip_address} is not found in any subnet, recovering...")
+                recovered.append(fip)
+                return
+            # get the floating IP reservation ID from neutron
+            try:
+                fip_info_from_neutron = fip_pool.show_floatingip(fip["floating_ip_address"])
+            except Exception as e:
+                # If the FIP is not registered in neutron that means there is no reservation for this IP
```
Comment (Mark-Powers) [inline blazar/plugins/floatingips/floatingip_plugin.py]: I think this should catch NeutronNotFound or whatever the specific exception is, and the comment should clarify something like:
"If the FIP is not created in neutron, then it is in the correct 'pending' state"
as it's not really about a "no reservation for this IP" but rather the IP is not already in use.

File: blazar/plugins/floatingips/floatingip_plugin.py [hunk 3]
```diff
@@ -447,3 +455,109 @@ def query_fip_allocations(self, fips, detail=None, lease_id=None,
...
+                LOG.debug("Error getting Floating IP from neutron - {e}, recovering...")
+                recovered.append(fip)
```
Comment (Mark-Powers) [inline blazar/plugins/floatingips/floatingip_plugin.py]: This shouldn't be added to recovered, as there should be nothing wrong with this case.

**Review by Mark-Powers — CHANGES_REQUESTED** [review]
This looks great! There are a few minor-ish concerns I have, but I think overall it would run fine.

I am not sure about the use of failed/recovered lists. My understanding based on the host monitor is these lists are meant to set an FIP as reservable/not reservable. That should be done only if neutron is misconfigured, or deleting the ip doesn't work. I think it is fine to add an already reservable ip to the "reservable" list, but that isn't done in the host monitor, and will probably generate a confusing log message like "123.123.123.123 recovered" for IPs that never had problems.

File: blazar/plugins/floatingips/floatingip_plugin.py (hunk 1 above)
Comment (msherman64) [reply blazar/plugins/floatingips/floatingip_plugin.py:512]: this try, except block looks good to me.
that exception indicates that the relevant ipv4 address is within a dhcp allocation pool on the public subnet.

**Review by msherman64 — APPROVED** [review]

**Review by Mark-Powers — APPROVED** [review]

**Pull request merged**